src/java/org/apache/cassandra/index/sai/StorageAttachedIndexGroup.java

```diff
@@ -0,0 +1,241 @@
+
+    StorageAttachedIndexGroup(ColumnFamilyStore baseCfs)
+    {
+        this.baseCfs = baseCfs;
+        this.queryMetrics = new TableQueryMetrics(baseCfs.metadata());
+        this.stateMetrics = new TableStateMetrics(baseCfs.metadata(), this);
+
+        Tracker tracker = baseCfs.getTracker();
+        tracker.subscribe(this);
+    }
+
+    @Nullable
+    public static StorageAttachedIndexGroup getIndexGroup(ColumnFamilyStore cfs)
+    {
+        return (StorageAttachedIndexGroup) cfs.indexManager.getIndexGroup(StorageAttachedIndexGroup.class);
+    }
+
+    @Override
+    public Set<Index> getIndexes()
+    {
+        return ImmutableSet.copyOf(indices);
+    }
+
+    @Override
+    public void addIndex(Index index)
+    {
+        assert index instanceof StorageAttachedIndex;
+        indices.add((StorageAttachedIndex) index);
+    }
+
+    @Override
+    public void removeIndex(Index index)
+    {
+        assert index instanceof StorageAttachedIndex;
+        boolean removed = indices.remove(index);
+        assert removed : "Cannot remove non-existing index " + index;
+        /*
+         * For the in-memory implementation we only need to unsubscribe from the tracker
+         */
+        if (indices.isEmpty())
+        {
+            baseCfs.getTracker().unsubscribe(this);
+        }
+    }
+
+    @Override
+    public void invalidate()
+    {
+        // in case of dropping table, sstable contexts should already been removed by SSTableListChangedNotification.
+        queryMetrics.release();
+        stateMetrics.release();
+        baseCfs.getTracker().unsubscribe(this);
+    }
+
+    @Override
+    public boolean containsIndex(Index index)
+    {
+        return index instanceof StorageAttachedIndex && indices.contains(index);
+    }
+
+    @Override
+    public Index.Indexer indexerFor(Predicate<Index> indexSelector,
+                                    DecoratedKey key,
+                                    RegularAndStaticColumns columns,
+                                    int nowInSec,
+                                    WriteContext ctx,
+                                    IndexTransaction.Type transactionType,
+                                    Memtable memtable)
+    {
+        final Set<Index.Indexer> indexers =
+                indices.stream().filter(indexSelector)
+                                .map(i -> i.indexerFor(key, columns, nowInSec, ctx, transactionType, memtable))
+                                .filter(Objects::nonNull)
+                                .collect(Collectors.toSet());
+
+        return indexers.isEmpty() ? null : new StorageAttachedIndex.IndexerAdapter()
+        {
+            @Override
+            public void insertRow(Row row)
+            {
+                forEach(indexer -> indexer.insertRow(row));
+            }
+
+            @Override
+            public void updateRow(Row oldRow, Row newRow)
+            {
+                forEach(indexer -> indexer.updateRow(oldRow, newRow));
```

Comment:
nit: We'd have to profile to know for sure, but I'm a little worried that this would create unnecessary `Action` objects when compared to...

```
for (Index.Indexer indexer : indexers)
    indexer.updateRow(oldRow, newRow);
```

(same for `insertRow()`)

Reply:
Agreed, I've no idea where that came from. I've got rid of it.